app/models/batch_script.rb

```diff
@@ -69,30 +79,70 @@ def stderr_path
     @stderr_path.blank? ? stdout_path : @stderr_path
   end
 
+  def attributes
+    {
+      arguments: nil, name: nil, stdout_path: nil, stderr_path: nil,
+    }
+  end
+
   # Must be called at the end of the job lifecycle to remove the script
   def cleanup
-    Async::IO::Threads.new.async do
-      FileUtils.rm_rf(File.dirname(path))
-    end.wait
+    if Async::Task.current?
+      Async::IO::Threads.new.async { FileUtils.rm_rf(dirname) }.wait
+    else
+      FileUtils.rm_rf(dirname)
+    end
   end
 
   def write
     Async::IO::Threads.new.async do
-      FileUtils.mkdir_p File.dirname(path)
-      File.write(path, content)
+      FileUtils.mkdir_p(dirname)
+      File.write(script_path, content)
       # We don't want the content hanging around in memory.
       self.content = nil
+
+      serialized_env = env.map { |k, v| "#{k}=#{v}" }.join("\0")
+      File.write(env_path, serialized_env)
+      # We don't want the env hanging around in memory.
+      self.env = nil
     end.wait
   end
 
   def content
     # We deliberately don't cache the value here.
-    @content || Async::IO::Threads.new.async { File.read(path) }.wait
+    return @content if @content
+    if Async::Task.current?
+      Async::IO::Threads.new.async { File.read(script_path) }.wait
+    else
+      File.read(script_path)
+    end
   rescue Errno::ENOENT
   end
 
-  def path
-    File.join(FlightScheduler.app.config.spool_dir, 'jobs', job.id.to_s, 'job-script')
+  def env
+    # We deliberately don't cache the value here.
+    return @env if @env
+    serialized_env = 
+      if Async::Task.current?
+        Async::IO::Threads.new.async { File.read(env_path) }.wait
+      else
+        File.read(env_path)
+      end
+    Hash[serialized_env.split("\0").map { |pairs| pairs.split('=', 2) }]
```

Comment:
Is there are particular reason why we are using the `\0` and `=` syntax for envs? I suspect that it is something to do with the `\n` line character.

However [ActiveMode::Serializer::JSON](https://api.rubyonrails.org/classes/ActiveModel/Serializers/JSON.html) should be able to encode newline characters perfectly fine, and makes serialisation easier.

Reply:
Probably not a good one.  It is pretty much what slurm uses and is what came to mind when thinking about how to do this.

Reply:
I'm really not sure if this or JSON would be better.  I think you might be right, but I have a nagging feeling that JSON has issues with storing arbitrary strings which I can't shake.  It would probably be good to revisit this later.

Reply:
JSON can't store binary data, hence why it occasionally needs to be used with `base64`. However it should be able to escape any arbitrary UNICODE string.

The current method wouldn't be able to encode binary data as it would probably contain `\0` characters.

Reply:
I've added a card to trello and this can be looked at at another time.